### PR TITLE
Add option to delete stop

### PR DESCRIPTION
**Pull request by Nikhil-Vats**
### Details

Adds the option to delete a stop using three dots menu with confirmation modal.

### Fixed Issues

$ https://github.com/Expensify/App/issues/25901
PROPOSAL: https://github.com/Expensify/App/issues/25901#issuecomment-1693818817

### Tests

1. Go to request money.
2. Go to the distance tab.
3. Click on add stop.
4. Click on three dots icon.
5. Verify that you see the option `Delete stop` there. Click on it.
6. Verify that you see the confirmation modal. Click on delete.
7. Verify that you are navgated back to previous screen with distance tab and the stop is deleted from the list of waypoints/stop.
- [x] Verify that no errors appear in the JS console

### Offline tests

Same as above.
### QA Steps

Same as above.
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/31413407/12c06b2e-4d64-43c1-ac1f-7d6f39935b21

https://github.com/Expensify/App/assets/31413407/585fe171-c149-42a9-82f3-88c5ffc6ec6f

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/31413407/2927972c-2a41-44bb-ac52-48012fdbf001

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/31413407/9c6b61f7-78d2-4faa-bf8b-5801b4f332ae

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/31413407/9d55abdc-a1b5-4d00-8d7c-b96828b2c21c

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/31413407/f9e21509-70b3-45ef-a88d-0aca54cdc2b9

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/31413407/bd691851-5cfc-4be4-be3a-a390eb175a86

</details>

### CODE REVIEW

**Comment by hayata-suenaga** [comment]
> The text used -

Let's use `waypoint` instead of `stop`. Please request Spanish translations for these 🙇 

> The behaviour of this feature for start and finish points -

When there are only two waypoints (start and finish), let's hide the three dot button

When there are several waypoints, even start or finish waypoints can be deleted. My only concern is that we're doing some weird stuff with how we store waypoints (I think waypoints are stored in an object with each key being `waypoint0` or `waypoint1`...

**Comment by hayata-suenaga** [comment]
> Can we also add the Waiting for Copy label on original GH right now or do we need to do it later?

thank you for your attention to the guideline. I really appreciate it. In this case, because the change is very simple, I don't think we need marketing approval for this. But we still need Spanish translation confirmation.

> This doesn't happen on staging so it is from a recent PR and it is also affecting our delete waypoint testing on mobile browsers.

so currently, when you first click the dot menu on the Waypoint Editor page, the menu popup doesn't appear?

**Comment by hayata-suenaga** [comment]
> should we wait for Spanish translation to add test proofs and merge PR?

you can record the screen or take screenshots before Spanish translation is approved. We should probably wait for Spanish translation before merging though

but again, we can do all the testing before Spanish

**Comment by hayata-suenaga** [comment]
> Can you please report this (if it hasn't been reported previously) and post a link for the report here?

@Nikhil-Vats by the way, did you have time to report the bug? (if you can find the PR that caused the regression, that's 💯

**Comment by hayata-suenaga** [comment]
>  @hayata-suenaga will decide if we should hold this until the iOS issue is fixed or not.

Looking at #expensify-bugs, the issue seems to be present on all popover menus 😭 let's NOT hold this issue and start test and merge 🚀

File: src/pages/iou/WaypointEditor.js
Comment (hayata-suenaga) [inline src/pages/iou/WaypointEditor.js:107]: can we rename this to `deleteStopAndHideModal`?

also, is it our a part of the style guideline that we cannot use name like `onConfirm` or something? (I know I'm the one who should know this 😭 )

File: src/pages/iou/WaypointEditor.js
Comment (hayata-suenaga) [inline src/pages/iou/WaypointEditor.js:137]: @Nikhil-Vats could you kindly teach me why we need this prop? Why does `HeaderWithBackButton` need to know the width of the screen?

File: src/pages/iou/WaypointEditor.js
Comment (hayata-suenaga) [reply src/pages/iou/WaypointEditor.js]: ah yep I remember it's from the checklist... anyway, thank you for the change and detailed explanations 🙇

File: src/pages/iou/WaypointEditor.js
Comment (hayata-suenaga) [reply src/pages/iou/WaypointEditor.js:137]: okay that makes sense thank you for your explanation 🙇

**Comment by hayata-suenaga** [comment]
@Nikhil-Vats which commit did you revert?

**Comment by hayata-suenaga** [comment]
@Nikhil-Vats I'm working on the fix for the popover behavior: https://github.com/Expensify/App/issues/25998

**Review by allroundexperts — COMMENTED** [review]
Great work here @Nikhil-Vats 🏅🏅🏅

Few minor comments and then we should be good here.

**Review by allroundexperts — APPROVED** [review]
Works well!

**Comment by hayata-suenaga** [comment]
> After deleting all the stops, if we haven't specified the start and end points, the map does not pan back to the original state (ie San Francisco). Instead, it remains on the deleted stop.

I'd say this is an expected behavior, but thank you for your attention to details

> You appear to be offline shows up twice in iOS.

Do you think this is a regression from Map related features?

**Comment by hayata-suenaga** [comment]
I think someone responded to the spanish translation question you posted could you recheck? @Nikhil-Vats ?

File: src/pages/iou/WaypointEditor.js
Comment (hayata-suenaga) [inline src/pages/iou/WaypointEditor.js:195]: Not a Blocker: you can de-structure here

```suggestion
        key: ({ transactionID }) => `${ONYXKEYS.COLLECTION.TRANSACTION}${transactionID}`,
```

**Review by allroundexperts — APPROVED** [review]

**Review by hayata-suenaga — DISMISSED** [review]
🚀

**Comment by hayata-suenaga** [comment]
> How can we proceed forward from here?

that's feature hasn't implemented yet 😓 It's something I'm working on

**Review by hayata-suenaga — DISMISSED** [review]

**Comment by hayata-suenaga** [comment]
oh no there is a conflict 😭 @Nikhil-Vats can you take care of it when you have time?

**Review by hayata-suenaga — APPROVED** [review]

**Comment by hayata-suenaga** [comment]
As [per this comment](https://github.com/Expensify/App/pull/25984#issuecomment-1696329012), we're merging the PR without lint test succeeding.

@Nikhil-Vats are you suggesting that [this PR](https://github.com/Expensify/App/pull/25974) caused the lint error?

**Comment by hayata-suenaga** [comment]
I didn't see your coment @Nikhil-Vats can you link it?

now I'm afraid that the lint error will fail tests on main 😨

**Comment by hayata-suenaga** [comment]
@allroundexperts yep let's open a PR

**Pull request merged**
